include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -420,124 +327,67 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
-                                             is_first_process_on_node);
-
-    // Continue as above but with the serial triangulation that gets
-    // distributed
-    unsigned int n_partitions = Utilities::MPI::n_mpi_processes(mpi_comm);
-    for(int level = tria_copy.n_global_levels() - 1; level >= 0; --level)
-    {
-      // reduce the number of MPI ranks per coarsening step by at most a
-      // factor of 8, in order to avoid too much transfer out of a single MPI
-      // process during the MG level transfer.
-      n_partitions = std::min(
-        n_partitions,
-        std::max(n_partitions / 8,
-                 std::max<unsigned int>(1U, serial_tria.n_active_cells() / n_cells_per_process)));
-
-      // extract relevant information from distributed triangulation
-      auto const construction_data =
-        TriangulationDescription::Utilities::create_description_from_triangulation_in_groups<dim,
-                                                                                             dim>(
-          [&](auto & tria) { tria.copy_triangulation(serial_tria); },
-          [&](auto & tria, auto const &, auto const) {
-#ifdef DEAL_II_WITH_METIS
-            GridTools::partition_triangulation(n_partitions, tria);
-#else
-            GridTools::partition_triangulation_zorder(n_partitions, tria);
-#endif
-          },
-          mpi_comm,
-          n_processes_per_node);
-
-      // create fully distributed triangulation
-      auto level_tria =
-        std::make_shared<parallel::fullydistributed::Triangulation<dim, spacedim>>(mpi_comm);
-
-      for(auto const i : fine_triangulation->get_manifold_ids())
-        if(i != numbers::flat_manifold_id)
-          level_tria->set_manifold(i, fine_triangulation->get_manifold(i));
-      level_tria->create_triangulation(construction_data);
-
-      // save mesh
-      coarse_grid_triangulations[level] = level_tria;
-
-      // coarsen mesh
-      if(is_first_process_on_node && level > 0)
-        serial_tria.coarsen_global();
-    }
+  unsigned int n_processes_last = Utilities::MPI::n_mpi_processes(mpi_comm);
+  // create coarse meshes
+  for(unsigned int l = max_level; l > 0; --l)
+  {
+    // coarsen mesh
+    temp_triangulation.coarsen_global();
+
+    // compute the partitions of the mesh
+
+    // TODO: the grain size in terms of the number of cells is currently
+    // hard-coded to 200 here, plus a factor to account for the fact that we
+    // do not want to reduce the number of participating MPI ranks by more
+    // than about a factor of 8, in order to limit the number of messages a
+    // process on the coarse level might need to handle.
+    unsigned int grain_size_limit =
+      8 * temp_triangulation.n_global_active_cells() / n_processes_last;
+    RepartitioningPolicyTools::MinimalGranularityPolicy<dim> policy(
+      std::min<unsigned int>(200, grain_size_limit));
+    auto const partition = policy.partition(temp_triangulation);
```

Comment:
I think `partition` is not an intuitive name here (I am also not a fan of using `auto` in such cases as it makes the code difficult to understand; I guess the problem also originates from naming in deal.II since it is not intuitively clear to me what `policy.partition()` is doing; `partition` can be understood as noun or verb here). Here, I think of `partition` as one part of the triangulation, but it is then unclear to me why `partition.linfty_norm()` further below has something to do with the number of processes? Further below, I would guess `partition` takes the role of an mpi comm?